stdlib/http/src/main/java/org/ballerinalang/net/http/websocket/WebSocketUtil.java

```diff
@@ -169,60 +178,168 @@ public static String getErrorMessage(Throwable err) {
 
         MapValue<String, Object> detailRecordType = BallerinaValues.createRecordValue(
                 packageName, WebSocketConstants.WEBSOCKET_ERROR_DETAILS);
         MapValue<String, Object> detailRecord = BallerinaValues.createRecord(detailRecordType, message, null);
         return BallerinaErrors.createError(reason, detailRecord);
     }
 
-    public static MapValue<String, Object> createDetailRecord(String errMsg) {
-        return createDetailRecord(errMsg, null);
+    /**
+     * Establish connection with the endpoint.
+     *
+     * @param webSocketClient - the WebSocket client.
+     * @param wsService - the WebSocket service.
+     */
+    public static void establishWebSocketConnection(ObjectValue webSocketClient, WebSocketService wsService) {
+        WebSocketClientConnectorListener clientConnectorListener = (WebSocketClientConnectorListener) webSocketClient.
+                getNativeData(WebSocketConstants.CLIENT_LISTENER);
+        WebSocketClientConnector clientConnector = (WebSocketClientConnector) webSocketClient.
+                getNativeData(WebSocketConstants.CLIENT_CONNECTOR);
+        boolean readyOnConnect = webSocketClient.getMapValue(CLIENT_ENDPOINT_CONFIG).getBooleanValue(
+                WebSocketConstants.CLIENT_READY_ON_CONNECT);
+        ClientHandshakeFuture handshakeFuture = clientConnector.connect();
+        handshakeFuture.setWebSocketConnectorListener(clientConnectorListener);
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        handshakeFuture.setClientHandshakeListener(new WebSocketClientHandshakeListener(webSocketClient, wsService,
+                clientConnectorListener, readyOnConnect, countDownLatch));
+        // Set the countDown latch for every handshake
+        waitForHandshake(webSocketClient, countDownLatch);
+    }
+
+    /**
+     * Check whether the client's config has the retryConfig property.
+     *
+     * @param webSocketClient - the WebSocket client.
+     * @return If the client's config has the retry config, then return true.
+     */
+    public static boolean hasRetryConfig(ObjectValue webSocketClient) {
+        return webSocketClient.getMapValue(CLIENT_ENDPOINT_CONFIG).
+                getMapValue(WebSocketConstants.RETRY_CONFIG) != null;
     }
 
-    public static MapValue<String, Object> createDetailRecord(String errMsg, ErrorValue cause) {
-        MapValue<String, Object> detail = BallerinaValues.createRecordValue(HttpConstants.PROTOCOL_HTTP_PKG_ID,
-                                                                            WebSocketConstants.WEBSOCKET_ERROR_DETAILS);
-        return BallerinaValues.createRecord(detail, errMsg, cause);
+    private static void waitForHandshake(ObjectValue webSocketClient, CountDownLatch countDownLatch) {
+
+        long timeout = WebSocketUtil.findTimeoutInSeconds((MapValue<String, Object>) webSocketClient.getMapValue(
```

Comment:
Shall we suppress warning?
```suggestion
               @SuppressWarnings(WebSocketConstants.UNCHECKED)
                long timeout = WebSocketUtil.findTimeoutInSeconds((MapValue<String, Object>) webSocketClient.getMapValue(
```